### PR TITLE
Feature: Support stlink v2 isol variant

### PR DESCRIPTION
**Pull request by HrMitrev**
## Detailed description
The [stlink-isol-v2](https://estore.st.com/en/st-link-v2-isol-cpn.html) variant differs from other stlink-v2 variants with how the SWDIO pin is multiplexed. This commit adds support for it. I have tested the commit on my STlink ISOL and it seems to work fine.
## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [ ] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues
N/A

### CODE REVIEW

File: src/platforms/stlink/stlink_common.c
Comment (dragonmux) [inline src/platforms/stlink/stlink_common.c:55]: This appears to be new pin setup code for clone models - could you explain briefly why this is required and correct?

File: src/platforms/stlink/platform.h
Comment (dragonmux) [inline src/platforms/stlink/platform.h:91]: Ideally this and the clear below would use the unpacked/inlined version of the code this calls, but this works. Just to clarify though - PA1 is the SWDIO direction setting pin? Want to be sure we've understood the purpose of this code.

**Review by dragonmux — COMMENTED** [review]
We've got two queries that could do with being answered, and then this looks like it's good to merge.

File: src/platforms/stlink/stlink_common.c
Comment (HrMitrev) [reply src/platforms/stlink/stlink_common.c:55]: It is not new, it was executed before at the end of the function(the if statement based on rev) and talking to @ALTracer it is used to force USB re enumeration after reset from the bootloader. I decided to delete the code which is used for rev detection and moved these 2 lines there

File: src/platforms/stlink/platform.h
Comment (HrMitrev) [reply src/platforms/stlink/platform.h:91]: Good point - I can inline the code

File: src/platforms/stlink/stlink_common.c
Comment (dragonmux) [reply src/platforms/stlink/stlink_common.c:55]: Ohhh, gotcha - ok, thank you for clarifying!

File: src/platforms/stlink/platform.c
Comment (ALTracer) [inline src/platforms/stlink/platform.c:63]: If you wanted to, you could implement feature parity with BMP v2.3 HW rev 6 w.r.t. floating the SWCLK pin, required to support stm32g0 in small pin-count packages. See Issue 945 and PR1192 (intentional not-a-mention), specifically commit https://github.com/blackmagic-debug/blackmagic/pull/1192/commits/a573c268d5cae2622c8bf659e608313c61d9801d for `native` platform.
PB12 is `T_SWDIO_IN` on **most/all** stlinkv2 schematics, and you made it a macro -- please substitute it into here (`SWDIO_IN_PIN`).
Accordingly, PA4 sounds like `TCK_DIR_PIN` and PA1 like `TMS_DIR_PIN`/`SWDIO_DIR_PIN`. This makes for less unused/magic pin numbers, even if you described their purpose in adjacent comments (nice!).

File: src/platforms/stlink/platform.c
Comment (ALTracer) [inline src/platforms/stlink/platform.c:86]: I wonder if that's true -- PB14 being T_JTMS (effectively T_SWDIO_OUT) was toggled to become input across BMF previously. Do you think PB14 forcing high/low SWDIO in no way interferes with receiving data on PB12 across the isolation circuitry? Although I don't have the exact schematics and don't know where the 100R between SWDIO_IN & SWDIO_OUT resides for ISOL PCB. But you seem to have the means to test it with live targets.

File: src/platforms/stlink/platform.h
Comment (ALTracer) [reply src/platforms/stlink/platform.h:91]: * Interesting when you put it that way. Does PA1 high *switch* the direction to input of levelshifter between T_SWDIO_OUT and PB14 coupler channel output, or *disconnect* it? I wonder if the shifter as input and coupler as output will fight (as the target drives its data against a fixed-level BMF TMS). In that case, consider also switching SWDIO_OUT to digital input mode. SWDIO_IN can and should stay.
* PA1 might be `TMS_DIR_PIN` (see the other comment). What about `TMS_SET_MODE()`? jtagtap will call it, is it identical in function to `SWDIO_MODE_DRIVE()`? I expect that after a SWD detach, your new platform may fail to drive JTAG TMS because of stuck direction pin. Please post logs with JTAG scan tests.

File: src/platforms/stlink/stlink_common.c
Comment (ALTracer) [reply src/platforms/stlink/stlink_common.c:55]: Indeed I gathered that this is a last-resort hack to present a long SE0 state on fixed 1.5k pullup boards to force re-enumeration. By the way, how do you expect to do it? Is PA15 USB_RENUM wired and available? Or this platform also has to perform the hack? (of inducing a babble (?) condition for lack of togglable pullup voltage source pin per USB 1.1/2.0 spec) Please confirm that the board works as BMF with ST_BOOTLOADER=1, unless you already wiped your stock bootloader with BMPBootloader/F1 (it may have the toggling because it calls `detect_rev()`)

**Review by ALTracer — COMMENTED** [review]
There are macro naming questions from me and a proposal to bring feature parity with SWCLK Hi-Z, which leads into considerations about drive contention somewhere on the PCB. I cannot "Review -1/+1" as I am not a maintainer, and I am very unlikely to get my hands on an ISOL, but you could test anything before the PR is merged, and other less knowledgable users started (hypothetical worst case scenario) damaging their boards. 
I assume you already live-tested your PR branch build on a unit?

File: src/platforms/stlink/platform.h
Comment (HrMitrev) [reply src/platforms/stlink/platform.h:91]: About the SWDIO multiplexing - the probe uses Si8420BB RF isolators which are fixed direction. If we ignore the isolators the schematic looks like this: 
image
PB14 and PA1 are always forced to output and PB12 is always forced to input because of the unidirectional RF isolators used to interface between the 2G125 IC and the STM32F01

SWDCLK should be similar, but I think PA4 is used to enable the output propagation or set it to high impedance

**Comment by HrMitrev** [comment]
> Ideally this and the clear below would use the unpacked/inlined version of the code this calls, but this works.

I have unpacked/inlined the mode switch of SWDIO and confirmed to be working on my isol probe. Also I have rebased. Currently I have used this patch for more than 2 weeks every day and my probe seems to going strong. I haven't had the chanse to confirm it is working with JTAG instead of SWD. In the future if I have more time and HW to test JTAG I will revisit. For now I think this can be merged

Cheers!

**Review by dragonmux — APPROVED** [review]
LGTM, thank you for tweaking and sorting that. We'll get this merged once the builds complete - thank you for the contribution!

File: src/platforms/stlink/platform.h
Comment (HrMitrev) [reply src/platforms/stlink/platform.h]: > About the SWDIO multiplexing - the probe uses Si8420BB RF isolators which are fixed direction. If we ignore the isolators the schematic looks like this: image PB14 and PA1 are always forced to output and PB12 is always forced to input because of the unidirectional RF isolators used to interface between the 2G125 IC and the STM32F01
> 
> SWDCLK should be similar, but I think PA4 is used to enable the output propagation or set it to high impedance

This was hidden after the changes, but may be a good idea to be easily accessible

**Comment by HrMitrev** [comment]
> About the SWDIO multiplexing - the probe uses Si8420BB RF isolators which are fixed direction. If we ignore the isolators the schematic looks like this: image PB14 and PA1 are always forced to output and PB12 is always forced to input because of the unidirectional RF isolators used to interface between the 2G125 IC and the STM32F01
> 
> SWDCLK should be similar, but I think PA4 is used to enable the output propagation or set it to high impedance

This was hidden after the changes, but may be a good idea to be easily accessible

**Pull request merged**
